_posts/2023-02-22-user-defined-operators.md

```diff
@@ -0,0 +1,484 @@
+    - You cannot define them to unnamed types using the type wildcard (`*`).
+1. Have operators that are not `pure`.
+    Operators cannot modify storage.
+    They can at most make pure external calls.
+
+    The worst that is technically possible with tricks is for them to modify memory (only through inline assembly)
+    or make staticcalls (through interfaces that do not match actual code deployed on chain).
+1. Have operators on mixed types.
+    Parameter and return value types of operators must be the same.
+    Also, no implicit conversions are possible, so it is very simple to determine if a particular operation
+    will invoke a user-define operator or a built-in one (just look at the type).
+1. Redefine operators.
+    You cannot replace any of the built-in operators with a custom definition.
+    For user-defined operators, only one definition can be provided.
+1. Bind operators to contract functions.
+    This also means that operator definitions cannot be inherited, virtual or remain unimplemented.
+1. Bind operators to overloaded functions.
+1. Use built-in functions to define operators.
+    You cannot e.g. bind an operator directly to `keccak256()`.
+1. Change precedence or associativity of an operator.
+1. Define any of the operators with more complex semantics.
+    You cannot define:
+    - `&&` and `||` due to their short-circuiting behavior.
+    - `++` and `--` because they modify their argument and also have separate prefix and postfix variants with different semantics.
+    - `+=`, `-=`, `^=`, `*=`, `/=`, `%=`, `&=`, `\|=`, `^=`, `<<=`, `>>=` because they modify their argument.
+    - `=`, `()`, `?:` and `.` because they are built-in for all types.
+    - `delete`, `new` and `[]` because they are not usable with UDVTs.
+1. Define custom operators that do not exist in the language, like `===` or unary `+`.
+1. Bind operators in bulk or accidentally.
+    The `using L for T` syntax for attaching a whole library to a type does not cover operators.
+    Operators must always be bound explicitly.
+
+In addition to these restrictions, we decided to limit them even more in response to feedback we received before releasing Solidity 0.8.18.
+This means you cannot also:
+1. Use library functions to define operators.
+    This excludes external, public and private functions.
+    It is also now impossible to apply modifiers to operators because free functions cannot have them.
+1. Have non-global operators, which has several consequenes:
+    - Operators can only be bound in the file that also defines the type.
+        They are available wherever the type is available and cannot have different definitions in different scopes.
+    - Local bindings at contract level are not possible (i.e. inside contracts and libraries).
+        This also makes it impossible to define operators for local types that are defined inside contracts or libraries.
+1. Define `!`, `<<`, `>>` and `**`.
```

Comment:
```suggestion
1. Have operators on mixed types.
    Parameter and return value types of operators must be the same.
    Also, no implicit conversions are possible, so it is very simple to determine if a particular operation
    will invoke a user-define operator or a built-in one (just look at the type).
1. Redefine operators.
    You cannot replace any of the built-in operators with a custom definition.
    For user-defined operators, only one definition can be provided.
1. Bind operators to contract functions.
    This also means that operator definitions cannot be inherited, virtual, or remain unimplemented.
1. Bind operators to overloaded functions.
1. Use built-in functions to define operators.
    You cannot e.g. bind an operator directly to `keccak256()`.
1. Change the precedence or associativity of an operator.
1. Define any of the operators with more complex semantics.
    You cannot define:
    - `&&` and `||` due to their short-circuiting behavior.
    - `++` and `--` because they modify their argument and also have separate prefix and postfix variants with different semantics.
    - `+=`, `-=`, `^=`, `*=`, `/=`, `%=`, `&=`, `\|=`, `^=`, `<<=`, `>>=` because they modify their argument.
    - `=`, `()`, `?:` and `.` because they are built-in for all types.
    - `delete`, `new`, and `[]` because they are not usable with UDVTs.
1. Define custom operators that do not exist in the language, like `===` or unary `+`.
1. Bind operators in bulk or accidentally.
    The `using L for T` syntax for attaching a whole library to a type does not cover operators.
    Operators must always be bound explicitly.

In addition to these restrictions, we decided to limit them even more in response to feedback we received before releasing Solidity 0.8.18.
This means you cannot also:
1. Use library functions to define operators.
    This excludes external, public, and private functions.
    It is also now impossible to apply modifiers to operators because free functions cannot have them.
1. Have non-global operators, which has several consequences:
    - Operators can only be bound in the file that also defines the type.
        They are available wherever the type is available and cannot have different definitions in different scopes.
    - Local bindings at the contract level are impossible (i.e. inside contracts and libraries).
        This also makes it impossible to define operators for local types that are defined inside contracts or libraries.
1. Define `!`, `<<`, `>>` and `**`.
```

Reply:
> the contract level

https://github.com/ethereum/solidity-blog/pull/202#discussion_r1113401728
> only allowed at file level

I'm getting contradictory suggestions here :) Which one is it?

`at contract level`/`at file level` sounds more correct to me because I'm not referring to any particular file or contract. I'm basically using it as an adjective.